### PR TITLE
Feature/tests

### PR DESCRIPTION
**Pull request by thecanadianroot**
**Warning :** Before creating a pull request, make sure that it does not already exists in the [pull requests](../). Thank you.

## Description
Give a brief description of the modifications
Added unit tests and general fixes + bags

## Fixes
Link all the related issues from the issue tracker
- Closes: #

## How has this been tested ?
(If possible) describe the steps to test these modifications.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,15 +1,12 @@
...
 
 USER root
 
-ARG BUILD_DATE
-ARG VERSION
-
 LABEL net.etsmtl.sonia-auv.node.build-date=${BUILD_DATE}
 LABEL net.etsmtl.sonia-auv.node.version=${VERSION}
-LABEL net.etsmtl.sonia-auv.node.name=${NODE_NAME}
+LABEL net.etsmtl.sonia-auv.node.name=proc_image_processing
```
Comment (ProgHunter) [inline Dockerfile:9]: not sure to understand this change?

File: src/proc_image_processing/cpu/algorithm/rot_rect.cc [hunk 2]
```diff
@@ -45,7 +45,7 @@ namespace proc_image_processing {
         angle = out_angle;
     }
 
-    void RotRect::swap(RotRect &a) {
+    void RotRect::swap(RotRect &a) noexcept {
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/algorithm/rot_rect.cc:48]: first time I see this keyword..

File: src/proc_image_processing/cpu/config.h [hunk 3]
```diff
@@ -9,12 +9,15 @@ namespace proc_image_processing {
 
     const std::string kProjectPath = std::string(std::getenv("NODE_PATH"));
 
-    const std::string kConfigPath = kProjectPath + "/config";
+    const std::string kConfigPath = kProjectPath + std::string(
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/config.h:12]: Good job on checking getenv() result! (else the constructor of the string throw an error, not always clear)

File: src/proc_image_processing/cpu/filters/custom/hide_submarine_frame_filter.h [hunk 4]
```diff
@@ -50,7 +55,7 @@ namespace proc_image_processing {
     private:
         RangedParameter<int> rotate_type_;
         cv::Mat bottom_mask_;
-        int prev_rot_value_;
+        int prev_rot_value_ = 0;
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/custom/hide_submarine_frame_filter.h:58]: nice safety

File: src/proc_image_processing/cpu/filters/custom/in_range_filter.h [hunk 5]
```diff
@@ -49,20 +49,22 @@ namespace proc_image_processing {
             cv::Mat hsv;
             cv::Mat luv;
 
+            // TODO
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/custom/in_range_filter.h]: todo what?

File: src/proc_image_processing/cpu/filters/custom/test_filter.h [hunk 6]
```diff
@@ -13,35 +13,48 @@ namespace proc_image_processing {
...
-                              specField2_.getValue());
+            target_.setTarget(
+                    header_.getValue(),
+                    x_.getValue() - 1000 / 2,
+                    y_.getValue(),
+                    w_.getValue(),
+                    h_.getValue(),
+                    angle_.getValue(),
+                    1000,
+                    -1000 - (1000 / 2),
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/custom/test_filter.h:40]: huh ? maybe use constants to clarify the numbers.. or just put -1500.

File: src/proc_image_processing/cpu/filters/filter_inl.h [hunk 7]
```diff
@@ -46,32 +55,8 @@ namespace proc_image_processing {
...
-        global_params_.addParameter(new RangedParameter<double>(name, value, min, max, &parameters_));
-    }
-
-    inline void Filter::addGlobalParameter(const std::string &name, const bool value) {
-        global_params_.addParameter(new Parameter<bool>(name, value, &parameters_));
-    }
-
-    inline void Filter::addGlobalParameter(const std::string &name, const std::string &value) {
-        global_params_.addParameter(new Parameter<std::string>(name, value, &parameters_));
+    inline const GlobalParameterHandler &Filter::getGlobalParamHandler() const {
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/filter_inl.h:58]: plus simple, mais ça brise un peu l'encapsulation des paramètres

File: src/proc_image_processing/cpu/filters/low_pass/clahe_filter.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                  clip_limit_("Clip limit", 40.0, &parameters_, "Threshold for contrast limiting."),
+                  tile_width_("Tile width", 8, &parameters_, "Width of tile used for histogram equalization."),
+                  tile_height_("Tile height", 8, &parameters_, "Height of tile used for histogram equalization.") {
+            setName("CLAHEFilter");
+        }
+
+        ~CLAHEFilter() override = default;
+
+        void apply(cv::Mat &image) override {
+            if (enable_()) {
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/low_pass/clahe_filter.h]: to delete

File: src/proc_image_processing/cpu/filters/low_pass/clahe_filter.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace proc_image_processing {
+
+    class CLAHEFilter : public Filter {
+    public:
+        using Ptr = std::shared_ptr<CLAHEFilter>;
+
+        explicit CLAHEFilter(const GlobalParameterHandler &globalParams)
+                : Filter(globalParams),
+                  enable_("Enable", false, &parameters_),
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/low_pass/clahe_filter.h]: to delete

File: src/proc_image_processing/cpu/filters/low_pass/clahe_filter.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                    dst.copyTo(lab_planes[0]);
+                    cv::merge(lab_planes, lab);
+
+                    cv::cvtColor(lab, image, CV_Lab2BGR);
+                }
+            }
+        }
+
+    private:
+        Parameter<bool> enable_, greyscale_;
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/low_pass/clahe_filter.h]: no need for enable_ here

File: src/proc_image_processing/cpu/filters/low_pass/white_filter.h [hunk 11]
```diff
@@ -14,24 +14,32 @@ namespace proc_image_processing {
...
         }
 
         ~WhiteFilter() override = default;
 
         void apply(cv::Mat &image) override {
-            cv::Scalar min_pixel_range = cv::Scalar(minimal_1pixel_range_(), minimal_2pixel_range_(),
-                                                    minimal_3pixel_range_());
-            cv::Scalar max_pixel_range = cv::Scalar(maximal_1pixel_range_(), maximal_2pixel_range_(),
-                                                    maximal_3pixel_range_());
+            cv::Mat mask;
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/filters/low_pass/white_filter.h]: ?

File: src/proc_image_processing/cpu/server/filter_chain.cc [hunk 12]
```diff
@@ -57,129 +56,156 @@ namespace proc_image_processing {
...
-        auto filter = Filter::Ptr(
-                FilterFactory::createInstance(filter_name,
-                                              param_handler_));
+    void FilterChain::addFilter(const std::string &name) {
+        auto filter = Filter::Ptr(FilterFactory::createInstance(name, param_handler_));
         if (filter != nullptr) {
             filters_.push_back(filter);
         } else {
-            throw std::invalid_argument("This filter does not exist in the library");
+            throw std::invalid_argument("Filter " + name + " does not exist!");
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/server/filter_chain.cc:208]: ty!

Comment (ProgHunter) [reply src/proc_image_processing/cpu/server/filter_chain.cc:208]: Great improvement of the error handling in this file!

File: Dockerfile (hunk 1 above)
Comment (thecanadianroot) [reply Dockerfile:9]: Environment variable NODE_NAME is not set anywhere, so the label value was essentially an empty one...

File: src/proc_image_processing/cpu/algorithm/rot_rect.cc (hunk 2 above)
Comment (thecanadianroot) [reply src/proc_image_processing/cpu/algorithm/rot_rect.cc:48]: Code smell (mais supposément qu'on doit définir ça explicitement si ça lance pas d'exceptions)

File: src/proc_image_processing/cpu/config.h (hunk 3 above)
Comment (thecanadianroot) [reply src/proc_image_processing/cpu/config.h:12]: Actually, it's to support using a different path for the tests

File: src/proc_image_processing/cpu/server/global_parameter_handler.h [hunk 13]
```diff
@@ -83,11 +83,11 @@ namespace proc_image_processing {
         static const char SEPARATOR = ';';
 
     private:
-        std::queue<Target> _vision_target;
+        std::queue<Target> target_queue_;
         // Using pointer here to preserve the object if
         // the vector is moved in memory.
-        std::vector<ParameterInterface *> _params_vec;
-        cv::Mat _original_image;
+        std::map<std::string, ParameterInterface *> params_;
```
Comment (ProgHunter) [inline src/proc_image_processing/cpu/server/global_parameter_handler.h:89]: nice

File: src/proc_image_processing/cpu/filters/custom/test_filter.h (hunk 6 above)
Comment (thecanadianroot) [reply src/proc_image_processing/cpu/filters/custom/test_filter.h:40]: Dunno, was already there. Just reformated the stuff

File: src/proc_image_processing/cpu/filters/filter_inl.h (hunk 7 above)
Comment (thecanadianroot) [reply src/proc_image_processing/cpu/filters/filter_inl.h:58]: C'est pas la job d'un filter d'ajouter des params globaux, ça devrait être le filter chain I think

File: test/black_box_test.cc [hunk 14]
```diff
@@ -128,9 +141,6 @@ TEST(BlackBoxTest, test) {
 
     // Try to end it again
     execute_service.call(executeCmdRequest, executeCmdResponse);
-
-    // test as much as you want here...
-    sleep(3600);
```
Comment (ProgHunter) [inline test/black_box_test.cc:133]: kek

File: Dockerfile (hunk 1 above)
Comment (ProgHunter) [reply Dockerfile:9]: haa, because the env is set just after in the file, ok

File: src/proc_image_processing/cpu/algorithm/rot_rect.cc (hunk 2 above)
Comment (ProgHunter) [reply src/proc_image_processing/cpu/algorithm/rot_rect.cc:48]: ok.. good to know I guess

File: src/proc_image_processing/cpu/filters/filter_inl.h (hunk 7 above)
Comment (ProgHunter) [reply src/proc_image_processing/cpu/filters/filter_inl.h:58]: yepp

**Review by ProgHunter — APPROVED** [review]
Good improvement of the code!

**Pull request merged**